Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/branch.py

```diff
@@ -90,20 +91,57 @@ def editable(cls, branch, repository=None):
         return True
 
     @classmethod
-    def branch_point(cls, repository, limit=None):
-        cnt = 0
-        commit = repository.commit(include_log=False, include_identifier=False)
-        while cls.editable(commit.branch, repository=repository):
-            cnt += 1
-            if limit and cnt > limit:
-                return None
-            commit = repository.find(argument='HEAD~{}'.format(cnt), include_log=False, include_identifier=False)
-            if cnt > 1 or commit.branch != repository.branch or cls.editable(commit.branch, repository=repository):
-                log.info('    Found {}...'.format(string_utils.pluralize(cnt, 'commit')))
-            else:
-                log.info('    No commits on editable branch')
+    def branch_point(cls, repository):
+        branches = repository.branches_for(remote=None)
+        production_branches = [
+            'remotes/{}/{}'.format(remote, branch)
+            for remote in repository.source_remotes()
+            for branch in branches[remote] if not repository.dev_branches.match(branch)
+        ]
+
+        head = run(
+            [repository.executable(), 'rev-parse', 'HEAD'],
+            cwd=repository.root_path,
+            capture_output=True,
+            encoding='utf-8',
+        ).stdout.strip()
+
+        partial_bases = set()
+        for shard in [
+            production_branches[cls.MERGE_BASE_SHARD_SIZE * i:cls.MERGE_BASE_SHARD_SIZE * (i + 1)]
+            for i in range(1 + len(production_branches) // cls.MERGE_BASE_SHARD_SIZE)
+        ]:
+            if not shard:
+                continue
+            result = run(
+                [repository.executable(), 'merge-base', head] + shard,
+                cwd=repository.root_path,
+                capture_output=True,
+                encoding='utf-8',
+            )
+            if result.returncode:
+                continue
```

Comment:
Why do we want to ignore all cases where this exits uncleanly? At the very least adding a comment here would be good.

Reply:
It's more of a "what else can we do?" situation. Callers of this function treat "None" as an error, so if we want to represent an error, that's what we should return. Which we could right here. But in a hypothetical situation where one shard fails but the others don't (and I'm not sure how you would even get into such a state), it's a good bet that one of the successful shards actually has the answer you're looking for.

Put another way: if all call to `merge-base` fail, we return `None` (which is the error state). If only one fails, what's the correct action? This code attempts to continue if a single shard fails.

Reply:
I think I'd rather just return `None` if any fail, because as you say it shouldn't happen, and from my point-of-view it's unclear what the result should be.

Reply:
I'm fine with that behavior, will modify